packages/packsquash/src/pack_file/commands_function_file.rs

```diff
@@ -0,0 +1,140 @@
+	type ByteChunkType = Vec<u8>;
+	type OptimizationError = OptimizationError;
+	type OptimizedByteChunksStream =
+		impl Stream<Item = OptimizedBytesChunk<Self::ByteChunkType, Self::OptimizationError>>;
+
+	fn process(self) -> Self::OptimizedByteChunksStream {
+		let mut line_number = LineNumber::new();
+
+		let minify = self.optimization_settings.minify;
+		let strip_bom = self.optimization_settings.strip_bom;
+
+		MarkLastDecorator::new(FramedRead::new(
+			self.read,
+			// Limit line length to 16 KiB to bound memory consumption and be nice to Minecraft.
+			// Longer lines are probably an error, and will negatively affect Minecraft performance
+			LinesCodec::new_with_max_length(16 * 1024)
+		))
+		.filter_map(move |(line_result, is_last)| {
+			let processed_line_result = line_result.map_or_else(
+				|err| Some(Err(err.into())),
+				|line| process_line(line, is_last, line_number, minify, strip_bom)
+			);
+
+			line_number.increment();
+
+			future::ready(processed_line_result)
+		})
+	}
+
+	fn is_compressed(&self) -> bool {
+		false
+	}
+}
+
+impl<T: AsyncRead + Send + Unpin + 'static> PackFileConstructor<T> for CommandsFunctionFile<T> {
+	type OptimizationSettings = CommandsFunctionFileOptions;
+
+	fn new(
+		file_read_producer: impl FnOnce() -> Option<AsyncReadAndSizeHint<T>>,
+		_: PackFileAssetType,
+		optimization_settings: Self::OptimizationSettings
+	) -> Option<Self> {
+		file_read_producer().map(|(read, _)| Self {
+			read,
+			optimization_settings
+		})
+	}
+}
+
+/// Processes the specified line of text, returning an optimized bytes chunk with its optimized
+/// representation. `None` is returned to signal that the line should not be copied to the output
+/// file at all.
+fn process_line<L: Into<String>>(
+	line: L,
+	is_last: bool,
+	line_number: LineNumber,
+	minify: bool,
+	strip_bom: bool
+) -> Option<OptimizedBytesChunk<Vec<u8>, OptimizationError>> {
+	const MINIFIED: &str = "Minified";
+	const NOT_MINIFIED: &str = "Copied";
+
+	let mut line = line.into();
+
+	// The lines codec takes care of stripping line breaks from the line, even if
+	// Windows line endings (CR + LF) are used. However, we should remove the BOM in
+	// the first line if present. This would remove the BOM prefix from the key of
+	// the first command string, avoiding pack author confusion and doing a better
+	// work than Mojang ;). This might break packs that dealt with the BOM by
+	// including it in the key references elsewhere, so only do it if the option
+	// is enabled
+	if line_number.is_first() && strip_bom && line.chars().next().map_or(false, |c| c == '\u{feff}') {
+		line.remove(0);
+	}
+
+	let trimmed_line = line.trim();
+
+	// Check whether the line is a comment. If so, bail out by copying or skipping
+	// it. It's copied only if we're not minifying
+	if trimmed_line.is_empty() || trimmed_line.starts_with('#') || trimmed_line.starts_with("//") {
+		(!minify).then(|| prepare_for_output(line, is_last, NOT_MINIFIED))
```

Comment:
`//` is never allowed in vanilla .mcfunction files from 1.12 onwards to begin a comment line: the game rejects any line that begins with a slash, but it gives a different error message when using a double slash.

Allowing `//` as a comment sequence has the unintended side effect of PackSquash outputting lines that might invalid to the game when minification is disabled, which is a bad thing. I see two ways of handling this:

- You take into account the pack metadata information to know what Minecraft version this pack targets, and decide whether to allow `//` or not based on that. This will work fine and not reject files that are valid for older versions, but it is more involved to implement and will allow pack authors to get away with arguably obsolete syntax in their functions, which will trade effort fixing functions now for effort fixing functions later, when they decide to upgrade the pack.
- Not allowing `//` altogether. This will reject valid files for older Minecraft versions, but due to the reasons stated above it's a bad idea to begin comments like this anyway. I like this option more, as it's simpler and encourages pack authors to future-proof their packs more, which is beneficial in the long term.

Reply:
How do you prefer to tackle this? :smile:

Reply:
If you use `//`, it will give a different error than `/`. This change is in this commit: e9b895630ed55b361d2ba72b83992c4f666bf245.